### PR TITLE
Bug: name prefix or suffix in base results in incorrect references

### PR DESCRIPTION
**Pull request by msk-**
Hopefully correctly produced per:
https://kubernetes-sigs.github.io/kustomize/contributing/bugs/

Follow-up to https://github.com/kubernetes-sigs/kustomize/issues/2609

### CODE REVIEW

**Comment by msk-** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> memo **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
>     * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
>     * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> 
>     * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> 
>     * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

A new comment

**Comment by msk-** [comment]
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[msk-](https://github.com/kubernetes-sigs/kustomize/pull/2829#)_
> To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **phanimarupaka**
> You can assign the PR to them by writing `/assign @phanimarupaka` in a comment when ready.
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=kubernetes-sigs%2Fkustomize).
> Needs approval from an approver in each of these files:
> 
>     * **[OWNERS](https://github.com/kubernetes-sigs/kustomize/blob/master/OWNERS)**
> 
> 
> Approvers can indicate their approval by writing `/approve` in a comment
> Approvers can cancel approval by writing `/approve cancel` in a comment

/assign @phanimarupaka

File: api/krusty/nameprefixsuffixpatch_test.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+	th.WriteF("/app/handlers/kustomization.yaml", `
+nameSuffix: -suffix
+
+resources:
+  - ./deployment.yaml
+`)
+
+	th.WriteF("/app/handlers/deployment.yaml", `
+apiVersion: extensions/v1beta1
```
Comment (monopole) [inline api/krusty/nameprefixsuffixpatch_test.go:29]:
```
apiVersion: apps/v1
```

File: api/krusty/nameprefixsuffixpatch_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+`)
+
+	th.WriteK("/app", `
+resources:
+  - ./mysql
+  - ./handlers
+
+configMapGenerator:
+  - name: mysql
+    behavior: merge
```
Comment (monopole) [inline api/krusty/nameprefixsuffixpatch_test.go:55]: This replaces _all the fields_ in the config map, with no "savings" in terms of configuration,
so the layout doesn't make sense in this isolated context.

So let's merge more variables in here to make it realistic.

Without merging in more vars, the immediate fix is this layout:
```
func TestNamePrefixSuffixPatchAlternative(t *testing.T) {
	th := kusttest_test.MakeHarness(t)

	th.WriteF("handlers/kustomization.yaml", `
nameSuffix: -suffix
resources:
- deployment.yaml
`)

	th.WriteF("handlers/deployment.yaml", `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: short
spec:
  template:
    spec:
      containers:
      - name: handler
`)

	th.WriteK(".", `
resources:
- handlers
configMapGenerator:
- name: mysql
  literals:
  - MYSQL_DATABASE=db
  - MYSQL_USER=user
patches:
- target:
    kind: Deployment
    name: s.*
  patch: |-
    kind: Deployment
    metadata:
      name: ignored
    spec:
      template:
        spec:
          containers:
            - name: handler
              envFrom:
                - configMapRef:
                    name: mysql
              env:
                - valueFrom:
                    configMapKeyRef:
                      name: mysql
                      key: MYSQL_DATABASE
`)

	m := th.Run(".", th.MakeDefaultOptions())
	th.AssertActualEqualsExpected(m, `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: short-suffix
spec:
  template:
    spec:
      containers:
      - env:
        - valueFrom:
            configMapKeyRef:
              key: MYSQL_DATABASE
              name: mysql-g9ttm44c48
        envFrom:
        - configMapRef:
            name: mysql-g9ttm44c48
        name: handler
---
apiVersion: v1
data:
  MYSQL_DATABASE: db
  MYSQL_USER: user
kind: ConfigMap
metadata:
  name: mysql-g9ttm44c48
`)
}
```

File: api/krusty/nameprefixsuffixpatch_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+	kusttest_test "sigs.k8s.io/kustomize/api/testutils/kusttest"
+)
+
+func TestNamePrefixSuffixPatch(t *testing.T) {
+	th := kusttest_test.MakeEnhancedHarness(t).
+		PrepBuiltin("PrefixSuffixTransformer").
+		PrepBuiltin("AnnotationsTransformer").
+		PrepBuiltin("LabelTransformer")
+	defer th.Reset()
```
[the remark this answers is not in the document]
Comment (msk-) [reply api/krusty/nameprefixsuffixpatch_test.go]: Ah, I tried removing 16-19, it worked, but I didn't actually know what I was doing so I left it as it was :grin:

File: api/krusty/nameprefixsuffixpatch_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                      name: mysql
+                env:
+                  - valueFrom:
+                      configMapKeyRef:
+                        name: mysql
+                        key: MYSQL_DATABASE
+`)
+
+	m := th.Run("/app", th.MakeDefaultOptions())
+	// I expected configMapRef.name and configMapKeyRef.name to be "mysql-g9ttm44c48" not "mysql"
```
[the remark this answers is not in the document]
Comment (msk-) [reply api/krusty/nameprefixsuffixpatch_test.go]: Yeah, I forget this is actually a PR, so it shouldn't have my lazy language in it.

File: api/krusty/nameprefixsuffixpatch_test.go (hunk 1 above)
Comment (msk-) [reply api/krusty/nameprefixsuffixpatch_test.go]: Living in the distant past here!

File: api/krusty/nameprefixsuffixpatch_test.go (hunk 2 above)
Comment (msk-) [reply api/krusty/nameprefixsuffixpatch_test.go]: I made it a little more realistic.

**Comment by msk-** [comment]
> Thanks so much for taking those bug filing instructions in the spirit intended.
> Saves a ton of time for everyone!

And thank you very much for the prompt response. I only wish I'd found that contribution page sooner. I wonder if it's possible to make it more visible when an issue is raised? (It's fine if I was dumb and missed it somewhere- that's on me).

I don't know if you're using the GH UI, but when you're making a comment on a code range, try pressing `ctrl+g` (or perhaps `cmd+g` on Mac) to make a code suggestion that the reviewee can accept with a button press. I guess it's less didactic to make the change _for_ the reviewee, but in the cases where you're writing out the replacements anyway, and they're trivial, it might make sense.

**Comment by monopole** [comment]
@msk thanks for the `ctrl-g` tip!  I have to start using that more so i remember it.


/lgtm

**Pull request merged**